src/math/GraphUtil.cpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <powsybl/math/GraphUtil.hpp>
+
+namespace powsybl {
+
+namespace math {
+
+GraphUtil::ConnectedComponentsComputationResult::ConnectedComponentsComputationResult(const std::vector<stdcxx::optional<unsigned long>>& componentNumber, const std::vector<unsigned long>& orderedComponents) :
+    m_componentNumber(componentNumber),
+    m_componentSize(orderedComponents) {
+}
+
+const std::vector<stdcxx::optional<unsigned long>>& GraphUtil::ConnectedComponentsComputationResult::getComponentNumber() const {
+    return m_componentNumber;
+}
+
+std::vector<stdcxx::optional<unsigned long >>& GraphUtil::ConnectedComponentsComputationResult::getComponentNumber() {
+    return m_componentNumber;
+}
+
+const std::vector<unsigned long>& GraphUtil::ConnectedComponentsComputationResult::getComponentSize() const {
+    return m_componentSize;
+}
+
+std::vector<unsigned long>& GraphUtil::ConnectedComponentsComputationResult::getComponentSize() {
+    return m_componentSize;
+}
+
+GraphUtil::ConnectedComponent::ConnectedComponent(unsigned long size) :
+    m_size(size) {
+}
+
+unsigned long GraphUtil::ConnectedComponent::getOrderedNumber() const {
+    return m_orderedNumber;
+}
+
+unsigned long GraphUtil::ConnectedComponent::getSize() const {
+    return m_size;
+}
+
+GraphUtil::ConnectedComponent& GraphUtil::ConnectedComponent::setOrderedNumber(unsigned long orderedNumber) {
+    m_orderedNumber = orderedNumber;
+    return *this;
+}
+
+GraphUtil::ConnectedComponentsComputationResult GraphUtil::computeConnectedComponents(const std::vector<std::vector<unsigned long>>& adjacencyList) {
+    std::vector<stdcxx::optional<unsigned long>> componentNumber(adjacencyList.size());
+    unsigned long c = 0UL;
+    std::vector<unsigned long> componentSize(adjacencyList.size(), 0UL);
+    for (unsigned long v = 0UL; v < adjacencyList.size(); v++) {
+        if (!componentNumber[v].is_initialized()) {
+            computeConnectedComponents(v, c, componentSize, adjacencyList, componentNumber);
+            c++;
+        }
+    }
+
+    // sort components by size
+    std::vector<std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>> components;
+    components.reserve(c);
+    std::vector<std::shared_ptr<std::pair<unsigned long, stdcxx::optional<unsigned long>>>> orderedComponents;
+    orderedComponents.reserve(c);
+
+    for (unsigned long i = 0UL; i < c; i++) {
+        auto comp = std::make_shared<std::pair<unsigned long, stdcxx::optional<unsigned long>>>(componentSize[i], stdcxx::optional<unsigned long>());
```

Comment:
why a shared_ptr?

Reply:
Because `components` and `orderedComponents` point on the same ConnectedComponents, but are ordered differently.